File: edukate-checker/src/main/java/io/github/sanyavertolet/edukate/checker/utils/CheckResultMessageUtils.java
package io.github.sanyavertolet.edukate.checker.utils;

import io.github.sanyavertolet.edukate.checker.dtos.ModelResponse;
import io.github.sanyavertolet.edukate.common.checks.CheckErrorType;
import io.github.sanyavertolet.edukate.common.checks.CheckResultMessage;
import io.github.sanyavertolet.edukate.common.checks.CheckStatus;
import io.github.sanyavertolet.edukate.common.checks.SubmissionContext;
import lombok.NonNull;
import lombok.experimental.UtilityClass;

@UtilityClass
public class CheckResultMessageUtils {

    public static CheckResultMessage success(
            @NonNull ModelResponse modelResponse,
            @NonNull SubmissionContext submissionContext
    ) {
        return new CheckResultMessage(
                submissionContext.getSubmissionId(),
                modelResponse.getStatus(),
                Math.max(
                        0f,
                        Math.min(1f, modelResponse.getTrustLevel())
                ),
                modelResponse.getStatus() == CheckStatus.SUCCESS
                        ? CheckErrorType.NONE
                        : modelResponse.getErrorType(),
                modelResponse.getExplanation()
        );
    }

    public static CheckResultMessage error(@NonNull SubmissionContext submissionContext) {
        return new CheckResultMessage(
                submissionContext.getSubmissionId(),
                CheckStatus.INTERNAL_ERROR,
                0f,
                CheckErrorType.NONE,
                "Automatic check failed. Please retry later or contact support."
        );
    }
}

File: edukate-checker/src/main/kotlin/io/github/sanyavertolet/edukate/checker/utils/CheckResultMessageUtils.kt
package io.github.sanyavertolet.edukate.checker.utils

import io.github.sanyavertolet.edukate.checker.dtos.ModelResponse
import io.github.sanyavertolet.edukate.common.checks.CheckErrorType
import io.github.sanyavertolet.edukate.common.checks.CheckResultMessage
import io.github.sanyavertolet.edukate.common.checks.CheckStatus
import io.github.sanyavertolet.edukate.common.checks.SubmissionContext

fun success(modelResponse: ModelResponse, submissionContext: SubmissionContext): CheckResultMessage =
    CheckResultMessage(
        submissionContext.submissionId,
        submissionContext.checkResultId,
        modelResponse.status,
        modelResponse.trustLevel.coerceIn(0f, 1f),
        if (modelResponse.status == CheckStatus.SUCCESS) CheckErrorType.NONE else modelResponse.errorType,
        modelResponse.explanation,
    )

fun error(submissionContext: SubmissionContext): CheckResultMessage =
    CheckResultMessage(
        submissionContext.submissionId,
        submissionContext.checkResultId,
        CheckStatus.INTERNAL_ERROR,
        0f,
        CheckErrorType.NONE,
        "Automatic check failed. Please retry later or contact support.",
    )
